clash-cores/src/Clash/Cores/Xilinx/DcFifo.hs

```diff
@@ -0,0 +1,271 @@
+    , Signal read (DataCount depth)
+    , Signal read a
+    )
+  go st@(FifoState _ rt) rdClk wrClk@(tWr :- _) =
+    if rt < fromIntegral tWr
+      then goRead st rdClk wrClk
+      else goWrite st rdClk wrClk
+
+  goWrite (FifoState _ rt) rdClk (tWr :- wrClk) rstR rEna (True :- rstWNext) (_ :- wData) =
+      (False :- preFull, False :- preOver, 0 :- preWCnt, fifoEmpty, under, rCnt, rData)
+    where
+      (preFull, preOver, preWCnt, fifoEmpty, under, rCnt, rData) =
+        go (FifoState mempty (rt-fromIntegral tWr)) rdClk wrClk rstR rEna rstWNext wData
+
+  goWrite (FifoState q rt) rdClk (tWr :- wrClk) rstR rEna (_ :- rstW) (wDat :- wDats1) =
+    (full, over, wCnt, fifoEmpty, under, rCnt, rData)
+    where
+      (preFull, preOver, preWCnt, fifoEmpty, under, rCnt, rData) =
+        go (FifoState q' (rt-fromIntegral tWr)) rdClk wrClk rstR rEna rstW wDats1
+
+      wCnt = sDepth q :- preWCnt
+      full = (Seq.length q == maxDepth) :- preFull
+      (q', over) =
+        if Seq.length q < maxDepth
+          then (case wDat of { Just x -> x Seq.<| q ; _ -> q }, False :- preOver)
+          else (q, isJust wDat :- preOver)
+
+  sDepth = fromIntegral . Seq.length
+
+  goRead (FifoState _ rt) (tR :- rdClk) wrClk (True :- rstRNext) (_ :- rEnas1) rstW wData =
+    (full, over, wCnt, fifoEmpty, under, rCnt, rData)
+    where
+      rData = deepErrorX "Reset" :- preRData
+      fifoEmpty = True :- preEmpty
+      rCnt = 0 :- preRCnt
+      under = False :- preUnder
+
+      (full, over, wCnt, preEmpty, preUnder, preRCnt, preRData) =
+        go (FifoState mempty (rt+fromIntegral tR)) rdClk wrClk rstRNext rEnas1 rstW wData
+
+  goRead (FifoState q rt) (tR :- rdClk) wrClk (_ :- rstRNext) (rEna :- rEnas1) rstW wData =
+    (full, over, wCnt, fifoEmpty, under, rCnt, rData)
+    where
+      rCnt = sDepth q :- preRCnt
+      fifoEmpty = (Seq.length q == 0) :- preEmpty
+      rData = nextData :- preRData
+
+      (full, over, wCnt, preEmpty, preUnder, preRCnt, preRData) =
+        go (FifoState q' (rt+fromIntegral tR)) rdClk wrClk rstRNext rEnas1 rstW wData
+
+      (q', nextData, under) =
+        if rEna
+          then
+            case Seq.viewr q of
+              Seq.EmptyR -> (q, deepErrorX "FIFO empty", True :- preUnder)
+              qData Seq.:> qDatum -> (qData, qDatum, False :- preUnder)
+          else (q, deepErrorX "Enable off", errorX "Enable off" :- preUnder)
+
+  initState :: FifoState a
+  initState = FifoState Seq.empty 0
+
+  wrClkSignal = case wClk of
+    Clock _ (Just wrPeriods) -> wrPeriods
+    Clock _ Nothing ->
+      let SDomainConfiguration _ (snatToNum -> period) _ _ _ _ = knownDomain @write in
+        pure period
+  rdClkSignal = case rClk of
+    Clock _ (Just rdPeriods) -> rdPeriods
+    Clock _ Nothing ->
+      let SDomainConfiguration _ (snatToNum -> period) _ _ _ _ = knownDomain @read in
+        pure period
+
+{-# NOINLINE dcFifo #-}
+{-# ANN dcFifo (InlineYamlPrimitive [minBound..maxBound] [i|
```

Comment:
```suggestion
{-# ANN dcFifo (InlineYamlPrimitive [minBound..] [i|
```

Reply:
Is this better? It's more concise but the former is more explicit.

Reply:
I don't really mind either way, but @DigitalBrains1 mentioned he wanted it this way.